pkg/controller/seed-controller-manager/constraint-controller/controller.go

```diff
@@ -0,0 +1,328 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// This controller syncs the kubermatic constraints to constraint on the user cluster.
+	ControllerName = "constraint_syncing_controller"
+	finalizer      = kubermaticapiv1.KubermaticUserClusterNsDefaultConstraintCleanupFinalizer
+	Key            = "default"
+	Value          = "true"
+	AddAction      = "add"
+	RemoveAction   = "remove"
+)
+
+type reconciler struct {
+	log                     *zap.SugaredLogger
+	workerNameLabelSelector labels.Selector
+	recorder                record.EventRecorder
+	namespace               string
+	seedClient              ctrlruntimeclient.Client
+}
+
+func opaPredicate() predicate.Funcs {
+	return kubermaticpred.Factory(func(o ctrlruntimeclient.Object) bool {
+		cluster, ok := o.(*kubermaticv1.Cluster)
+		if !ok {
+			return false
+		}
+		return cluster.Spec.OPAIntegration != nil && cluster.Spec.OPAIntegration.Enabled
+	})
+}
+
+func withEventFilter() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			oldConstraint, ok := e.ObjectOld.(*kubermaticv1.Constraint)
+			if !ok {
+				return false
+			}
+			newConstraint, ok := e.ObjectNew.(*kubermaticv1.Constraint)
+			if !ok {
+				return false
+			}
+			return !reflect.DeepEqual(oldConstraint.Spec, newConstraint.Spec)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+}
+
+func Add(ctx context.Context,
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+	workerName string,
+	namespace string,
+	numWorkers int) error {
+
+	workerSelector, err := workerlabel.LabelSelector(workerName)
+	if err != nil {
+		return fmt.Errorf("failed to build worker-name selector: %v", err)
+	}
+
+	reconciler := &reconciler{
+		log:                     log.Named(ControllerName),
+		workerNameLabelSelector: workerSelector,
+		recorder:                mgr.GetEventRecorderFor(ControllerName),
+		namespace:               namespace,
+		seedClient:              mgr.GetClient(),
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Cluster{}},
+		enqueueConstraints(reconciler.seedClient, reconciler.log, namespace),
+		workerlabel.Predicates(workerName),
+		opaPredicate(),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for clusters: %w", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Constraint{}},
+		&handler.EnqueueRequestForObject{},
+		kubermaticpred.ByNamespace(namespace),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for seed constraints: %v", err)
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.Constraint{}},
+		&handler.EnqueueRequestForObject{},
```

Comment:
Here we should not enqueue the object itself, but the owner, isn't it?

Reply:
makes sense, on it!